Paper: NeuroConv. Streamlining Neurophysiology Data Conversion to the NWB Standard (#1088)
scipy-conference/scipy_proceedings: papers/mayorquin_neuroconv/main.md

---
# Ensure that this title is the same as the one in `myst.yml`
title: NeuroConv. Streamlining Neurophysiology Data Conversion to the NWB Standard
keywords: Neurodata Without Borders, NWB, Neurophysiology, Data standardization, Data conversion, DANDI, Python, Scientific software, Large-scale data
abstract: |

>   Converting diverse neurophysiology data to the standardized Neurodata Without Borders (NWB) format remains a significant barrier to data sharing and reuse. We present NeuroConv, a software that enables converting 44 distinct data formats while handling high-volume data as well as extracting meaningful metadata. The library has enabled the standardization of over 300 datasets totaling more than 300 TB in the DANDI archive and has been deployed to create automated conversion pipelines for 40 laboratories with unique experimental paradigms. By reducing technical barriers to NWB adoption, our tools accelerate progress toward reproducible neuroscience research through standardized data sharing.

Reviewer: ```suggestion
  Converting diverse neurophysiology data to the standardized Neurodata Without Borders (NWB) format remains a significant barrier to data sharing and reuse. We present NeuroConv, a software that enables converting 44 distinct data formats while handling high-volume data as well as extracting meaningful metadata. The library has enabled the standardization of over 350 datasets totaling more than 350 TB in the DANDI archive and has been deployed to create automated conversion pipelines for 40 laboratories with unique experimental paradigms. By reducing technical barriers to NWB adoption, our tools accelerate progress toward reproducible neuroscience research through standardized data sharing.
```

## NeuroConv Architecture and Design

To address these multifaceted challenges [NeuroConv](https://neuroconv.readthedocs.io/en/stable/index.html), a library that automates the ingestion and conversion of neurophysiology data from diverse formats into NWB. The development of NeuroConv required solving the following fundamental challenges:

```{list-table} Key technical challenges addressed by NeuroConv's architecture, highlighting the multifaceted requirements for automated neurophysiology data conversion.
:label: tbl:neuroconv-challenges
:header-rows: 1
* - Challenge
  - Solution Approach
* - Handling format and metadata diversity
  - Abstracting the complexity of 44+ distinct data formats while preserving format-specific metadata and ensuring NWB compliance
* - Managing high-volume data efficiently  
  - Processing datasets that exceed available RAM through streaming and chunked operations
* - Supporting complex experimental setups

>   - Accommodating multi-stream conversions where different modalities are recorded simultaneously in different formats

Reviewer 2: ```suggestion
  - Accommodating multi-stream conversions where different modalities are recorded simultaneously in different formats, while preserving time synchronization
```

```

The following sections detail how NeuroConv's architecture addresses each of these challenges through a modular, extensible design that maintains both flexibility and ease of use.

### Handling Diverse Data Formats

The challenge of format diversity in neurophysiology extends beyond their sheer number. Many formats, such as Neuralynx, exist in multiple versions, while others, like TIFF, exhibit significant internal variability in how labs use them. NeuroConv addresses this complexity through a modular architecture built around DataInterface classes. Each supported format has a dedicated DataInterface that handles data and metadata extraction, with specialized implementations like SpikeGLXRecordingInterface for raw voltage recordings, PhySortingInterface for spike-sorted data, and DeepLabCutInterface for behavioral tracking data. The critical contribution is a common interface that abstracts the format internal details and allows users to build conversions in a consistent manner, regardless of source format.

> NeuroConv addresses format diversity through a unified architecture centered on the DataInterface abstraction. Each supported format has a dedicated DataInterface that encapsulates the format-specific logic for data reading, metadata extraction, and NWB conversion while presenting a consistent API to users. The DataInterface serves as the fundamental building block of NeuroConv, providing a standardized pathway from diverse source formats to NWB output. This abstraction enables users to work with any supported format using identical code patterns, regardless of the underlying format complexity. The minimal conversion pipeline follows a three-step process illustrated in {ref}`fig:assets/minimal_conversion_pipeline`:

Reviewer 2: this reads a bit redundant:

```
The critical contribution is a common interface that abstracts the format internal details and allows users to build 
conversions in a consistent manner, regardless of source format.

NeuroConv addresses format diversity through a unified architecture centered on the DataInterface abstraction. Each 
supported format has a dedicated DataInterface that encapsulates the format-specific logic for data reading, metadata 
extraction, and NWB conversion while presenting a consistent API to users. The DataInterface serves as the 
fundamental building block of NeuroConv, providing a standardized pathway from diverse source formats to NWB 
output.
```

Author: That's true.

Author: Addressed, combined those two paragraphs into one.

> NeuroConv addresses format diversity through a unified architecture centered on the DataInterface abstraction. Each supported format has a dedicated DataInterface that encapsulates the format-specific logic for data reading, metadata extraction, and NWB conversion while presenting a consistent API to users. The DataInterface serves as the fundamental building block of NeuroConv, providing a standardized pathway from diverse source formats to NWB output. This abstraction enables users to work with any supported format using identical code patterns, regardless of the underlying format complexity. The minimal conversion pipeline follows a three-step process illustrated in {ref}`fig:assets/minimal_conversion_pipeline`:

Reviewer 2: ```
The minimal conversion pipeline follows a three-step process illustrated in...
```

the figure doesn't really look like a 3-step process, one can't tell what these steps are just looking at it... maybe we could have bold titles or something highlighting the steps as they are presented further down the text: 
1. instantiate the appropriate DataInterface
2. extract and optionally modify metadata
3. execute the conversion

Author: This is a good idea. I think I will change the text in this one because modifying  the figure is way more effort.

Author: Done.

:::{figure} assets/minimal_conversion_pipeline.png
:label: fig:assets/minimal_conversion_pipeline
The process begins with source data (e.g., binary recordings, metadata, and configuration files). A data-specific Interface object is instantiated and used to extract metadata via the .get_metadata() method. The resulting metadata can be optionally edited by the user to fill in missing or corrected fields. The finalized metadata and source data are then passed to the .run_conversion() method, which writes a complete NWB file compliant with the standard.
:::

Programmatically, this process can be summarized in a few lines of code, as shown below. The DataInterface handles all format-specific complexities internally, from parsing proprietary binary structures to extracting embedded metadata, while ensuring the output adheres to NWB best practices:

```python
from neuroconv import DataInterface

# Initialize the DataInterface for a specific format
interface = DataInterface(file_path="path/to/data/file")
# Extract the metadata from the source format
metadata = interface.get_metadata()
# 
# Code to modify the metadata as needed

> # 

Reviewer 2: we could illustrate the second step here, instead of having just an empty comment line:
```suggestion
metadata["NWBFile"]["experimenter"] = [Baggins, Bilbo"]
```

Author: Good idea.

# Add the data to an NWB File and write it to disk 
interface.run_conversion(nwbfile_path="path/to/nwbfile.nwb", metadata=metadata)
```

This core pattern remains consistent across all supported formats: instantiate the appropriate DataInterface, extract and optionally modify metadata, then execute the conversion. The DataInterface handles all format-specific complexities internally, from parsing proprietary binary structures to extracting embedded metadata, while ensuring the output adheres to NWB best practices.

> Currently supporting 44 distinct input formats (Table 1), each DataInterface is comprehensively documented and demonstrated in the [Conversion Gallery](https://neuroconv.readthedocs.io/en/stable/conversion_examples_gallery/index.html), where users can find complete examples requiring only ~5 lines of code to perform full data conversion. Throughout the conversion process, NeuroConv enforces NWB Best Practices for metadata and data organization while at the same time optimizes data storage for both archival purposes and cloud computing requirements.

Reviewer 2: ```suggestion
Currently supporting 44 distinct input formats (Table 1), each DataInterface is comprehensively documented and demonstrated in the [Conversion Gallery](https://neuroconv.readthedocs.io/en/stable/conversion_examples_gallery/index.html), where users can find complete examples requiring only ~5 lines of code to perform full data conversion. Throughout the conversion process, NeuroConv enforces NWB Best Practices for metadata and data organization while optimizing data storage for both archival purposes and cloud computing requirements.
```

### Handling Multi Stream Conversions

Neurophysiology experiments typically involve multiple simultaneous data streams from different modalities, such as raw electrophysiology recordings, spike-sorted data, and behavioral video. Each stream may be recorded in a different format, leading to complex conversion requirements. NeuroConv's architecture supports multi-stream conversions through the aggregation of DataInterfaces within a Converter framework as illustrated {ref}`fig:assets/diagram_converter`

:::{figure} assets/diagram_converter.png
:label: fig:assets/diagram_converter
The Converter orchestrates multiple specialized interfaces (A, B, C), each handling different data types (electrophysiology, imaging, and behavior). Individual interfaces extract metadata from their respective data sources, which the Converter combines into a single, user-editable metadata structure through its get_metadata() method. The Converter's run_conversion() method then coordinates all interfaces to produce a unified NWB file containing all data modalities. This design pattern enables flexible, modular integration of heterogeneous neuroscience data into a single standardized format.
:::

> The converter pattern enables combining multiple DataInterface instances into a single conversion workflow. This allows users to convert all relevant data streams from an experiment into a single NWB file, ensuring that all data is properly aligned and associated with the correct metadata. The Converter class handles the orchestration of multiple DataInterfaces, managing the order of operations and resolving any conflicts in metadata or data organization.

Reviewer 2: ```suggestion
The converter pattern enables combining multiple DataInterface instances into a single conversion workflow. This allows users to convert all relevant data streams from an experiment into a single NWB file, ensuring that all data is properly aligned and associated with the correct metadata. The Converter class handles the orchestration of multiple DataInterfaces, managing the order of operations and resolving any conflicts in metadata or data organization. Here's an example of conversion for a multi-modal experimental session:
```

```python
from neuroconv import ConverterPipe
from neuroconv.datainterfaces import SpikeGLXRecordingInterface, PhySortingInterface, DeepLabCutInterface

# Initialize the DataInterfaces for each data stream
recording_interface = SpikeGLXRecordingInterface(file_path="path/to/recording/file")
sorting_interface = PhySortingInterface(file_path="path/to/sorting/file")
behavior_interface = DeepLabCutInterface(file_path="path/to/behavior/file")

# Create the ConverterPipe with the DataInterfaces
data_interfaces = [recording_interface, sorting_interface, behavior_interface]
converter = ConverterPipe(
    data_interfaces=data_interfaces
)

metadata = converter.get_metadata()
# Modify metadata as needed

> 

Reviewer 2: like above, it would be nice with an example:
```suggestion
metadata["NWBFile"]["experimenter"] = [Baggins, Bilbo"]
```

A critical feature is the ability to process datasets larger than available RAM. NeuroConv inherits from the work performed by the NWB core group with [iterative writing](https://pynwb.readthedocs.io/en/stable/tutorials/advanced_io/plot_iterative_write.html#sphx-glr-tutorials-advanced-io-plot-iterative-write-py) to stream data in manageable chunks, with configurable chunk sizes based on available resources. We have extended this approach to support chunked reading from SpikeInterface, enabling buffered processing of numerous extracellular electrophysiology formats, such as SpikeGLX, Neuralynx, and Plexon. We have also implemented an iterative writing approach for roiextractors that allows buffered writing of large imaging datasets, such as those generated by whole-brain calcium imaging. Furthermore, we have implemented chunked solutions for other data-intensive formats such as video. This approach enables processing of arbitrarily large files and has been successfully tested on 100+ GB files using computers with only 8 GB of RAM.

> For storage optimization, NeuroConv leverages HDF5 and Zarr's support for chunked, compressed datasets. The current supported backends in NWB are HDF5 and Zarr. Compression algorithms represent a trade-off between storage space and access speed [@alessio_compression_2023]. NeuroConv exposes an easy-to-use [API](https://neuroconv.readthedocs.io/en/stable/user_guide/backend_configuration.html) for configuring chunking and compression settings at the dataset level, allowing for quick experimentation while providing sensible defaults that work for most users.

Reviewer 2: ```suggestion
For storage optimization, NeuroConv leverages HDF5 and Zarr's -- the current supported backends in NWB -- support for chunked, compressed datasets. Compression algorithms represent a trade-off between storage space and access speed [@alessio_compression_2023]. NeuroConv exposes an easy-to-use [API](https://neuroconv.readthedocs.io/en/stable/user_guide/backend_configuration.html) for configuring chunking and compression settings at the dataset level, allowing for quick experimentation while providing sensible defaults that work for most users.
```

Determining optimal chunk parameters involves complex tradeoffs [@zarr_performance; @nguyen2023impact]. Large chunks minimize the number of read operations but may require decompressing unnecessary data when accessing small subsets. Small chunks provide more precise access but increase overhead, particularly for cloud storage where each chunk requires a separate HTTP range request. Generally, appropriate chunking requires understanding the most common data access patterns. Since neurophysiology has relatively standardized analysis workflows and visualization patterns, it becomes feasible to implement evidence-based heuristics for chunk sizing across common data types, such as voltage recordings and imaging datasets.

> 

Reviewer 2: ```suggestion
### Multi-modal time synchronization  

Precise temporal alignment across diverse recording modalities is essential for accurate multi-modal data analysis and reproducibility. NeuroConv streamlines this critical process by providing intuitive, unified methods for time synchronization, leveraging common temporal references like hardware clocks or synchronization pulses. In many cases, it automatically detects and reconciles temporal offsets between devices, reducing manual effort and ensuring NWB files maintain internally consistent timestamps across all modalities. This automation enforces best practices for temporal metadata in NWB and enhances downstream analysis integrity.
```

:::

At the foundational level, PyNWB and MatNWB provide direct programmatic access to the NWB specification, offering maximum flexibility but requiring deep understanding of the standard. NWB GUIDE offers a graphical interface that democratizes access to NWB conversion through guided workflows, making it accessible to users without programming experience. NeuroConv complements both approaches by providing programmatic automation while maintaining the flexibility needed for complex, multi-modal experimental setups.

This ecosystem approach ensures that researchers can choose the conversion method that best matches their technical expertise and experimental complexity, while all approaches converge on the same standardized output format.

### Integration with Data Archives and Visualization

> The Distributed Archives for [Neurophysiology Data Integration (DANDI)](https://dandiarchive.org/) platform complements NWB by providing free hosting for NWB-formatted datasets up to terabytes in size. DANDI offers researchers a pathway to meet NIH data sharing requirements while effectively archiving their data and leveraging an expanding ecosystem of visualization and analysis tools. The platform supports versioned datasets, comprehensive metadata, and API access, making it an ideal complement to NeuroConv's conversion capabilities.

Reviewer 2: ```suggestion
The [Distributed Archives for Neurophysiology Data Integration (DANDI)](https://dandiarchive.org/) platform complements NWB by providing free hosting for NWB-formatted datasets up to terabytes in size. DANDI offers researchers a pathway to meet NIH data sharing requirements while effectively archiving their data and leveraging an expanding ecosystem of visualization and analysis tools. The platform supports versioned datasets, comprehensive metadata, and API access, making it an ideal complement to NeuroConv's conversion capabilities.
```

Author: Thanks.

* **NWB Standard Evolution**: We actively participate in NWB schema development and maintain close alignment with emerging standards through engagement with [NWB Extensions Proposals](https://github.com/nwb-extensions/nwbep-review/). By monitoring and contributing to current developments—including enhanced schemas for experimental events (NWBEP001), extracellular electrophysiology (NWBEP002), and optical physiology (NWBEP003, NWBEP004)—we ensure that NeuroConv incorporates the latest standard improvements as they become available for the benefit of our users.
* **AI-Assisted Conversion**: We are exploring large language model integration to advance our core mission of automating neurophysiology data conversion. This includes using LLMs to generate DataInterfaces from source format documentation, automatically extract metadata from experimental protocols, and generate custom conversion pipelines based on natural language requirements (conversion agent).

> * **Cloud-Optimized Storage**: We are implementing improved chunking patterns and compression strategies for large datasets to enhance cloud access performance. This includes systematic experimentation through the [NWB Benchmarks Project](https://nwb-benchmarks.readthedocs.io/en/latest/) to determine optimal chunk sizes and compression algorithms [@alessio_compression_2023]. Our goal is, on one hand, to implement evidence-based heuristics that ensure efficient and performant data storage and in the other as well as having a friendly and clear API that allows customization and flexibility for users to adapt to their specific needs. In the future, we would also like to explore the inclusions of cost optimization considerations for cloud storage [@cost_effective_scientific_datasets_2016].

Reviewer 2: ```suggestion
* **Cloud-Optimized Storage**: We are implementing improved chunking patterns and compression strategies for large datasets to enhance cloud access performance. This includes systematic experimentation through the [NWB Benchmarks Project](https://nwb-benchmarks.readthedocs.io/en/latest/) to determine optimal chunk sizes and compression algorithms [@alessio_compression_2023]. Our goal is to implement evidence-based heuristics that ensure efficient and performant data storage, as well as having a friendly and clear API that allows customization and flexibility for users to adapt to their specific needs. In the future, we would also like to explore the inclusions of cost optimization considerations for cloud storage [@cost_effective_scientific_datasets_2016].
```

## Closing Remarks

> We believe that NeuroConv represents a critical step toward realizing the vision of FAIR neurophysiology data. By automating the conversion of diverse data formats into a common standard, we enable researchers to focus on scientific discovery rather than data wrangling. The success of this approach depends not only on technical implementation but on fostering a community that values standardization, reproducibility, and open science.

Reviewer 2: ```suggestion
We believe that NeuroConv represents a critical step toward realizing the vision of FAIR (Findable, Accessible, Interoperable, and Reusable) neurophysiology data. By automating the conversion of diverse data formats into a common standard, we enable researchers to focus on scientific discovery rather than data wrangling. The success of this approach depends not only on technical implementation but on fostering a community that values standardization, reproducibility, and open science.
```

Our work demonstrates that effective scientific software development requires balancing automation with flexibility, standardization with customization, and ease of use with powerful capabilities. The challenges we have addressed such as format diversity, metadata complexity, and scale are not unique to neurophysiology but represent broader issues in scientific computing that require community-driven solutions.

The broader implications extend beyond neurophysiology to any scientific domain grappling with data heterogeneity and the need for standardization. We believe that our approach of abstracting format complexity through unified interfaces, while maintaining extensibility through modular architecture, provides a template for similar challenges in other fields.

> As the neurophysiology community continues to generate increasingly complex and voluminous datasets, tools like NeuroConv become essential infrastructure for scientific progress. By lowering barriers to data standardization and sharing, we contribute to a future where scientific data is truly FAIR (findable, accessible, interoperable, and reusable) accelerating discovery and enhancing reproducibility across the field.

Reviewer 2: This acronym appears first up there, so I moved the description there:
```suggestion
As the neurophysiology community continues to generate increasingly complex and voluminous datasets, tools like NeuroConv become essential infrastructure for scientific progress. By lowering barriers to data standardization and sharing, we contribute to a future where scientific data is truly FAIR accelerating discovery and enhancing reproducibility across the field.
```

Author: Yes!

---
# Ensure that this title is the same as the one in `myst.yml`
title: "NeuroConv: Streamlining Neurophysiology Data Conversion to the NWB Standard" 
keywords: Neurodata Without Borders, NWB, Neurophysiology, Data standardization, Data conversion, DANDI, Python, Scientific software, Large-scale data
abstract: |

>   Converting diverse neurophysiology data to the standardized Neurodata Without Borders (NWB) format remains a significant barrier to data sharing and reuse. We present NeuroConv, a software that enables converting 47 distinct data formats while handling high-volume data as well as extracting meaningful metadata. The library has enabled the standardization of over 390 datasets totaling more than 350 TB in the DANDI archive and has been deployed to create automated conversion pipelines for 40 laboratories with unique experimental paradigms. By reducing technical barriers to NWB adoption, our tools accelerate progress toward reproducible neuroscience research through standardized data sharing.

Reviewer: I don't think we can claim to have enabled every dataset on DANDI. I also don't know if these stats are that important for the abstract

Reviewer: How about this:

Modern neurophysiology generates increasingly complex, multimodal datasets that require standardized formats for effective sharing and reuse. The Neurodata Without Borders (NWB) format has emerged as a leading solution for data standardization, but conversion from diverse proprietary formats remains a significant bottleneck due to format heterogeneity, metadata complexity, and the specialized expertise required for implementation. We present NeuroConv, an open-source Python library that automates the conversion of neurophysiology data from 47 distinct formats into NWB through a unified, modular architecture. NeuroConv addresses key challenges through format-specific DataInterfaces that abstract parsing complexity while maintaining a consistent API, multi-stream Converters that integrate heterogeneous data modalities, and optimized handling of large-scale datasets through chunked processing and cloud-compatible storage strategies. The library leverages established parsing libraries from the broader neuroscience software ecosystem while contributing improvements upstream, ensuring robust format support and community sustainability. NeuroConv's design enables researchers to convert complex, multi-modal experimental sessions with minimal code while preserving critical metadata and temporal alignment across recording systems. By democratizing access to data standardization, NeuroConv removes technical barriers that have historically limited NWB adoption, enabling laboratories regardless of technical expertise to contribute to a growing ecosystem of standardized, reusable neurophysiology data. This work represents a critical step toward FAIR (Findable, Accessible, Interoperable, and Reusable) neurophysiology data, accelerating scientific discovery through enhanced data sharing and reproducibility.

Author: Added directly from the word document.

## Integration with the Neuroscience and Wider Software Ecosystem

NeuroConv pursues a “no-wheel-reinvention” strategy: wherever a mature open-source library already parses a file format or implements a preprocessing step, NeuroConv delegates that responsibility instead of duplicating it. This design choice concentrates development effort on the NWB hand-off layer while reducing maintenance burden and maximising compatibility with community standards.

For electrophysiology workflows, NeuroConv leverages NEO [@neo] for format parsing and SpikeInterface [@spikeinterface] for unified access to raw and spike-sorted data, with ProbeInterface [@probeinterface] handling probe geometry metadata. This hierarchical approach enables NeuroConv to inherit comprehensive support for acquisition systems spanning from cutting-edge Neuropixels arrays to legacy multi-electrode platforms, while ensuring that preprocessing pipelines, spike sorting results, and quality metrics from SpikeInterface integrate seamlessly into the final NWB representation.

> For optical physiology, NeuroConv maintains [ROIExtractors](https://github.com/catalystneuro/roiextractors), which provides a unified API for optical physiology data across both raw imaging and processed segmentation outputs. For raw imaging data, ROIExtractors relies heavily on TiffFile [@doi:10.5281/zenodo.6795860] to decode the heterogeneous TIFF variants produced by modern microscopy systems, handling diverse acquisition formats and metadata structures. For segmentation data, ROIExtractors integrates outputs from Suite2P [@suite2p], CaImAn [@caiman], EXTRACT [@extract_2017; @extract_2021], and CNMF-E [@cnmfe_e_2016, @cnmf_e_2018] analysis pipelines. This hierarchical approach ensures consistent NWB representation of raw image stacks, ROI masks, fluorescence traces, and deconvolved neural activity signals across different processing workflows.

Reviewer 3: ```suggestion
For optical physiology, the NeuroConv team maintains [ROIExtractors](https://github.com/catalystneuro/roiextractors), which provides a unified API for optical physiology data across both raw imaging and processed segmentation outputs. For raw imaging data, ROIExtractors relies heavily on TiffFile [@doi:10.5281/zenodo.6795860] to decode the heterogeneous TIFF variants produced by modern microscopy systems, handling diverse acquisition formats and metadata structures. For segmentation data, ROIExtractors integrates outputs from Suite2P [@suite2p], CaImAn [@caiman], EXTRACT [@extract_2017; @extract_2021], and CNMF-E [@cnmfe_e_2016, @cnmf_e_2018] analysis pipelines. This hierarchical approach ensures consistent NWB representation of raw image stacks, ROI masks, fluorescence traces, and deconvolved neural activity signals across different processing workflows.
```

Author: Section moved up but I changed this. Thanks.

> Behavioral data integration is considerably more heterogeneous in nature and therefore resists a unified API such as the one for electro and optical physiology described previously. To handle this heterogeneity, NeuroConv leverages multiple specialized libraries to handle diverse data types. Pose estimation trajectories are processed through dedicated Python APIs: [SLEAP-IO](https://github.com/talmolab/sleap-io) for SLEAP [@sleap_2022] data, custom adapters for DeepLabCut [@deep_lab_cut_2018] and Lightning Pose [@lightning_pose_2024] outputs. Audio waveforms from auditory experimental setups are processed using SciPy [@scipy], while video streams for behavioral video streams are handled through OpenCV [@bradski2000opencv], which delegates codec operations to [FFmpeg](https://ffmpeg.org/). Additionally, we utilize [PyMatReader](https://pymatreader.readthedocs.io/en/latest/) for MATLAB .mat files, which remain prevalent in neuroscience for storing experimental data and analysis results. This library provides a robust interface for MATLAB file formats, enabling seamless integration of experimental data into NWB files.

Reviewer 3: ```suggestion
Behavioral data integration is considerably more heterogeneous in nature and therefore resists a unified API such as the one for electro and optical physiology described previously. To handle this heterogeneity, NeuroConv leverages multiple specialized libraries to handle diverse data types. Pose estimation trajectories are processed through dedicated Python APIs: [SLEAP-IO](https://github.com/talmolab/sleap-io) for SLEAP [@sleap_2022] data, custom adapters for DeepLabCut [@deep_lab_cut_2018] and Lightning Pose [@lightning_pose_2024] outputs. Audio waveforms from auditory experimental setups are processed using SciPy [@scipy], while video streams for behavioral video streams are handled through OpenCV [@bradski2000opencv], which delegates codec operations to [FFmpeg](https://ffmpeg.org/). Additionally, we utilize [PyMatReader](https://pymatreader.readthedocs.io/en/latest/) for MATLAB `.mat` files, which remain prevalent in neuroscience for storing experimental data and analysis results. This library provides a robust interface for MATLAB file formats, enabling seamless integration of experimental data into NWB files.
```

Reviewer 3: Lol, the GitHub change UX for this is pretty dismal - I'm just suggesting tickmarks around ".mat" suffix

Though I'd point out you might actually want to remove the ".mat" statement here entirely since the other formats don't list out possible corresponding suffixes either

Also, ".mat" is only convention, not requirement, since such a data file can have and be recognized by any suffix (since it is internal header values that specify it to be readable as MATLAB format)

Author: Yes, I think google docs works better for this because the unit of the diff is the word and not the line.

I also removed this section. I moved it up in the last commit and the suffix is not there. 

Thanks.

# Current Limitations

While NeuroConv has significantly improved data standardization processes, some major challenges remain:

**Format Coverage**: Despite supporting 47 formats, the neurophysiology ecosystem continues to evolve with new acquisition systems, format versions, and processing software. Each new system often introduces proprietary formats with unique metadata structures and data organization schemes. While NeuroConv's modular architecture allows users to develop custom DataInterfaces for unsupported formats, this process requires substantial technical expertise in both the source format's internal structure and NeuroConv's interface architecture. Additionally, maintaining custom interfaces as both the source format and NeuroConv evolve presents ongoing maintenance challenges for individual laboratories.

> **Custom Laboratory Formats**: Many laboratories develop custom data storage solutions tailored to their specific experimental workflows, often utilizing MATLAB .mat files, custom CSV schemas, or bespoke binary formats. These custom formats frequently exhibit significant variability not only between laboratories but even within the same laboratory over time as experimental protocols evolve. The heterogeneous nature of these formats—ranging from simple tabular data to complex nested structures with laboratory-specific metadata conventions—makes automated conversion particularly challenging. While NeuroConv can handle some standardized custom formats, the diversity of these approaches often requires manual intervention or custom code development. We recommend that laboratories work with data in its original acquisition format whenever possible, as this provides the richest metadata and most reliable conversion pathway.

Reviewer 3: ```suggestion
**Custom Laboratory Formats**: Many laboratories develop custom data storage solutions tailored to their specific experimental workflows, often utilizing MATLAB files, custom CSV schemas, or bespoke binary formats. These custom formats frequently exhibit significant variability not only between laboratories but even within the same laboratory over time as experimental protocols evolve. The heterogeneous nature of these formats—ranging from simple tabular data to complex nested structures with laboratory-specific metadata conventions—makes automated conversion particularly challenging. While NeuroConv can handle some standardized custom formats, the diversity of these approaches often requires manual intervention or custom code development. We recommend that laboratories work with data in its original acquisition format whenever possible, as this provides the richest metadata and most reliable conversion pathway.
```

Whether or not you keep the `.mat` suffix in the section above, for styling I'd suggest removing it here to be consistent with the two mentioned after

Author: Yes.

---
# Ensure that this title is the same as the one in `myst.yml`

> title: NeuroConv. Streamlining Neurophysiology Data Conversion to the NWB Standard

Author: @scipy-conference/2025-proceedings

I am unable to use `:` (colon) in the title. Is there a way around this or is that a limitation of the format?

Author: @rowanc1 

Hi, sorry, is the above the right tag to get help from the team? I am taking it from the youtube presentation but is not marked as a proper handle.

Reviewer 4: You can put it in quotes!

```suggestion
title: "NeuroConv: Streamlining Neurophysiology Data Conversion to the NWB Standard"
```

Author: Thanks!

Also, I asked below if there is a way to add citations in an HTML table or if there is a better way than my workaround.

Author: Thanks, this worked as well!

---

# Introduction, Background and Motivation
Modern neurophysiology produces complex, multimodal, large-scale datasets that require disciplined organization, shareability, and long-term archiving. The [Neurodata Without Borders (NWB)](https://nwb.org/) format has become a leading solution, embedding rich contextual metadata directly alongside primary data to preserve interpretability over time [@nwb_2015; @nwb_2022]. Crucially, NWB can unite disparate modalities such as simultaneous behavioral and electrophysiological recordings typically stored in separate files and formats, thereby enabling rigorous re-analysis and bolstering reproducibility.

> Yet several factors still hinder broad NWB adoption. Researchers must first grasp NWB’s detailed organizational schema; a steep learning curve for many laboratories. Using the core APIs, [pyNWB](https://pynwb.readthedocs.io/en/stable/) and [MatNWB](https://nwb.org/matnwb/), also demands coding proficiency that many researchers lack, placing the burden on a handful of technically skilled team members.

Reviewer 3: ```suggestion
Yet several factors still hinder broad NWB adoption. Researchers must first grasp NWB’s detailed organizational schema; a steep learning curve for many laboratories. Using the core APIs, [PyNWB](https://pynwb.readthedocs.io/en/stable/) and [MatNWB](https://nwb.org/matnwb/), also demands coding proficiency that many researchers lack, placing the burden on a handful of technically skilled team members.
```

Going by the capitalization form seen in the [README](https://github.com/NeurodataWithoutBorders/pynwb?tab=readme-ov-file#latest-release) (and the way I've always known it)

> Additional obstacles arise from the intrinsic diversity of neurophysiology. Modalities range from optical microscopy to extracellular and intracellular electrophysiology and varied behavioral-tracking techniques, each with specialized requirements [@nwb_2015; @MEF3_format_2016; @nwb_2022]. Dozens of acquisition systems record in proprietary formats optimized for rapid data capture, not for storage efficiency or cross-platform use. Consequently, formats differ widely in efficiency, longevity of support, metadata depth, and interoperability.

Reviewer 3: ```suggestion
Additional obstacles arise from the intrinsic diversity of neurophysiology. Modalities range from optical microscopy to extracellular and intracellular electrophysiology and varied behavioral-tracking techniques, each with specialized requirements [@nwb_2015; @MEF3_format_2016; @nwb_2022]. Dozens of acquisition systems record in proprietary formats optimized for rapid data capture, not for storage efficiency or cross-platform use. Consequently, formats differ widely in efficiency, longevity of support, metadata depth, and interoperability. For these reasons, many data archives in the field have chosen only to accept transformed and standardized data formats such as NWB. 
```

Maybe something like this, also emphasizing the goal (via archives) is to share the data

Author: I like it.

Author: But I am unsure about the part that says "only".

Reviewer 3: Hence why I end with 'such as' since they do accept other standards, but only a few blatantly accept uncurated data dumps (like CRCNS)

Reviewer 3: For example, SPARC might accept raw data formats if they are common, simple, and have enough metadata for someone to reasonably read it

Author: Thank you, I was not aware of this. But this is a great suggestion to improve the logical workflow of those sections. As in, providing reasons for "this is why automated conversions are important"

> Conversion therefore becomes the main bottleneck. Automated pipelines are essential to prevent repetitive effort and to guarantee that future datasets enter NWB seamlessly, but building such pipelines is hard. A single lab might combine voltage recordings, optical imaging, optogenetics, and behavioral tracking, demanding expertise in experimental design, source formats, and NWB itself, expertise rarely concentrated in one person or even one lab.

Reviewer 3: I feel like the jump from the previous statement to this lacks context - one can maybe piece this together in their mind, but stating something concrete could help the flow

Author: Let me put some thought into this.

:::{figure} assets/minimal_conversion_pipeline.png
:label: fig:assets/minimal_conversion_pipeline
The process begins with source data (e.g., binary recordings, metadata, and configuration files). A data-specific DataInterface object is instantiated and used to extract metadata via the .get_metadata() method. The resulting metadata can be optionally edited by the user to fill in missing or corrected fields. The finalized metadata and source data are then passed to the .run_conversion() method, which writes a complete NWB file compliant with the standard.
:::

> Programmatically, this process can be summarized in a few lines of code, as shown below. The DataInterface handles all format-specific complexities internally, from parsing proprietary binary structures to extracting embedded metadata, while ensuring the output adheres to NWB best practices:

Reviewer 3: ```suggestion
Programmatically, this process can be summarized in a few lines of pseudocode, as shown below. The DataInterface handles all format-specific complexities internally, from parsing proprietary binary structures to extracting embedded metadata, while ensuring the output adheres to NWB best practices:
```

Just pointing out that what is show below seems as if you could simply copy/paste it, perhaps change a path or two, and it should run

But that is not the case; `DataInterface` isn't a real import from the main package, the Shire isn't a valid input for `ZoneInfo`, and nested output paths will likely run into problems if the parent directories don't yet exist

But I get this example is purely to demonstrate 'the way a snippet usually looks'

Author: Yes, I am hoping the Shire gives away that this is not running code and is only for illustration purposes but maybe I can add an additional sentence to make it extra clear?

Reviewer 3: The single word 'psueodcode' is clear that it is not meant to run

But the last code segment (see other comments) is so close to being able to run as it is if you became consistent with this user metadata

Author: OK,

```python
from datetime import datetime
from zoneinfo import ZoneInfo
from neuroconv import DataInterface

# Initialize the DataInterface for a specific format
interface = DataInterface(file_path="path/to/data/file")

# Extract the metadata from the source format
metadata = interface.get_metadata()

# Modify metadata as needed, add missing or correct existing fields

> metadata["NWBFile"]["experimenter"] = ["Baggins, Bilbo"]
> metadata["NWBFile"]["experiment_description"] = "Example neurophysiology experiment"
> metadata["NWBFile"]["institution"] = "University of Middle Earth"
> # session_start_time is required for conversion
> metadata["NWBFile"]["session_start_time"] = datetime(2020, 1, 1, 12, 30, 0, tzinfo=ZoneInfo("Middle-earth/Shire"))

Reviewer 3: ```suggestion
metadata["NWBFile"]["experimenter"] = ["Doe, Jane", "Doe, John"]
metadata["NWBFile"]["experiment_description"] = "Example neurophysiology experiment."
metadata["NWBFile"]["institution"] = "Center for Open Neuroscience"
# session_start_time is required for conversion
metadata["NWBFile"]["session_start_time"] = datetime(2020, 1, 1, 12, 30, 0, tzinfo=ZoneInfo("US/Pacific"))
```

Another consideration could be to use a more standard John/Jane Doe 'somewhere' in the real world

While this wouldn't change the 'pseudocode' pattern for this specific block, doing it below in https://github.com/scipy-conference/scipy_proceedings/pull/1088/files#diff-cf1ba15a0431f36db0da839cd7c0ba464a306de5a9dca7c5e759fe441b631f96R263-R267 would also get that section close to being able to run 'as-is', if someone were to download some example SpikeGLX data and set the paths

<tr style="background-color: #F5F9FF"><td style="padding: 8px;">Neuralynx NVT</td></tr>
<tr style="background-color: #E8F2FF"><td style="padding: 8px;">SLEAP<sup><a href="#sleap-ref">6</a></sup></td></tr>
<tr style="background-color: #D6E7FF">
<td style="padding: 8px; color: #2E5090;"><em>Audio/Video</em></td>
<td style="padding: 8px;">Videos</td>
</tr>
<tr style="background-color: #F5F9FF">
<td style="padding: 8px; color: #2E5090;"><em>Operant Conditioning</em></td>
<td style="padding: 8px;">MedPC</td>
</tr>

<!-- General Data -->
<tr style="border-top: 3px solid #2E5090; background-color: #E8F2FF">
<td rowspan="4" style="padding: 8px; border-right: 2px solid #B8D4F1;"><strong>General Data</strong></td>
<td style="padding: 8px; color: #2E5090;"><em>Image</em></td>
<td style="padding: 8px;">Image (png, jpeg, tiff, etc)</td>
</tr>
<tr style="background-color: #D6E7FF">
<td rowspan="3" style="padding: 8px; color: #2E5090;"><em>Text/Tabular</em></td>
<td style="padding: 8px;">CSV</td>
</tr>
<tr style="background-color: #E2EFFF"><td style="padding: 8px;">Excel</td></tr>
<tr style="background-color: #D6E7FF"><td style="padding: 8px;">Text</td></tr>
</tbody>
</table>

**References:**

> <a id="caiman-ref">1.</a> CaImAn [@caiman]  

Author: @fwkoch I am struggling to be able to do citations inside this table. The table is HTML and I want the names of some formats to be a citation. The workaround that I am using now is a reference and adding the citation on here as a footnote. 

Is there any other workaround? Also, sorry for quoting your directly. The tag I got from the video does not seem to work.

Reviewer 4: ```
<tr style="background-color: #E8F2FF"><td style="padding: 8px;">SLEAP<sup><cite id="caiman">6</cite></sup></td></tr>
```

You can use the `<cite id="caiman">` to do a hover citation directly.

Is that what you are looking for?

If you leave the node blank as well, it should fill in the citation text/style as well.

Author: Thank you! This is exactly what I wanted!

Author: ![image](https://github.com/user-attachments/assets/12747ef8-3fe6-4338-b647-7e2daccc9a20)

Reviewer 4: Amazing. :)

```python
from datetime import datetime
from zoneinfo import ZoneInfo
from pathlib import Path
from neuroconv.datainterfaces import IntanRecordingInterface

file_path = f"{ECEPHY_DATA_PATH}/intan/intan_rhd_test_1.rhd" # This can also be .rhs
interface = IntanRecordingInterface(file_path=file_path, verbose=False)

# Extract what metadata we can from the source files
metadata = interface.get_metadata()
# session_start_time is required for conversion. If it cannot be inferred
# automatically from the source files you must supply one.
session_start_time = datetime(2020, 1, 1, 12, 30, 0, tzinfo=ZoneInfo("US/Pacific"))
metadata["NWBFile"].update(session_start_time=session_start_time)

nwbfile_path = f"{path_to_save_nwbfile}"  # This should be something like: "./saved_file.nwb"
interface.run_conversion(nwbfile_path=nwbfile_path, metadata=metadata)
```

> As NeuroConv's format support has expanded to 47+ formats, dependency management has become increasingly complex. Each format often requires specialized libraries with potentially conflicting version requirements, creating dependency resolution challenges that can make installation difficult or impossible. For example, libraries for reading different formats may require different versions of Python or different versions of common dependencies such as numpy. Additionally, installing the dependencies of every DataInterface would create an unnecessary and inefficient environment with hundreds of packages, many of which users never need for their conversion.

Reviewer 3: ```suggestion
As NeuroConv's format support has expanded to 47+ formats, dependency management has become increasingly complex. Each format often requires specialized libraries with potentially conflicting version requirements, creating dependency resolution challenges that can make installation difficult or impossible. For example, libraries for reading different formats may require different versions of Python or different versions of common dependencies such as NumPy. Additionally, installing the dependencies of every DataInterface would create an unnecessary and inefficient environment with hundreds of packages, many of which users never need for their conversion.
```

Following NumPy capitalization pattern from [their docs](https://numpy.org/doc/stable/index.html#numpy-documentation)

Author: Thanks.

```python
pip install "neuroconv[spikeglx,phy,deeplabcut]"
```

This approach aggregates only the required dependencies for selected formats, avoiding dependency conflicts while ensuring that the installation remains lightweight and manageable for end users. Users working with a specific subset of formats can maintain clean, minimal environments without the overhead of unused dependencies.

## Multi Stream Conversions

Neurophysiology experiments typically involve multiple simultaneous data streams from different modalities, such as raw electrophysiology recordings, spike-sorted data, and behavioral video. Each stream may be stored in a different format, leading to complex conversion requirements. NeuroConv's architecture supports multi-stream conversions through the aggregation of DataInterfaces within a Converter framework as illustrated {ref}`fig:assets/diagram_converter`

:::{figure} assets/diagram_converter.png
:label: fig:assets/diagram_converter

> The Converter orchestrates multiple specialized interfaces (A, B, C), each handling different data types (electrophysiology, imaging, and behavior). Individual interfaces extract metadata from their respective data sources, which the Converter combines into a single, user-editable metadata structure through its get_metadata() method. The Converter's run_conversion() method then coordinates all interfaces to produce a unified NWB file containing all data modalities. This design pattern enables flexible, modular integration of heterogeneous neuroscience data into a single standardized format.

Reviewer 3: Just noting some inconsistency across the label text with regards to code invokations

https://github.com/scipy-conference/scipy_proceedings/pull/1088/files#diff-cf1ba15a0431f36db0da839cd7c0ba464a306de5a9dca7c5e759fe441b631f96R40 prefaces each with a `.`, as in `.get_metadata()`

None wrap in simple ` for code-style (does that work within the labels?)

Author: I don't understand, are you saying some `get_metadata` have the dot at the beginning and others do not?

Sorry, maybe I am slowt today.

Reviewer 3: I'm not positing either style 😉

Just noting the two label texts are inconsistent in that way, so whichever style you prefer I'd just make it the same in both

My opinion would be to do full markdown-style if the paper supports it - that is, "\`.get_metadata()\`" to more clearly differentiate code from text (and attract the eye)

Author: Yes, I agree. They should be consistent and I will try the code backticks to see if it works that way. Thanks.

Author: Fixed it, unified it to `.get_metadata`

> acknowledgments: |

Reviewer 5: Suggestion nit: While this may be a requirement of the template, it's slightly unconventional to have the acknowledgments at the very beginning of the paper. If the style guide allows, consider moving this section to the end, just before the references, which is a more traditional placement. If it's fixed by the template, please disregard this comment.

Author: In the preview it appears at the end:

![image](https://github.com/user-attachments/assets/5114c1d3-97dc-49b1-a316-5e0719be12b2)

Is this what you mean?

<td style="padding: 8px; font-weight: 600;">Scale challenges</td>
<td style="padding: 8px;">Dataset sizes frequently reach hundreds of gigabytes to terabytes, requiring specialized handling for memory-efficient processing</td>
</tr>
<tr style="background-color: rgba(27, 87, 136, 0.08)">
<td style="padding: 8px; font-weight: 600;">Multi-modal integration</td>
<td style="padding: 8px;">Experiments often combine multiple recording systems (electrophysiology, imaging, behavior) each storing data in different formats. This also introduces the problem of data alignment and synchronization across modalities.</td>
</tr>
</tbody>
</table>
:::

# NeuroConv: Architecture and Design

To address the challenges described in {ref}`sec:introduction` (see {ref}`tbl:nwb-challenges`), we developed [NeuroConv](https://neuroconv.readthedocs.io/en/stable/index.html), an open source library that automates the ingestion and conversion of neurophysiology data from diverse formats into NWB. This section describes in detail how NeuroConv's architecture addresses each of these challenges through a modular, extensible design that maintains both flexibility and ease of use.

## Handling Format Diversity

> The challenge of format diversity in neurophysiology extends beyond their sheer number (47 supported at the moment). Many formats, such as Neuralynx, have multiple versions, while others, like TIFF, exhibit significant internal variability in how labs use them. NeuroConv addresses this complexity through a unified architecture centered on the DataInterface abstraction. DataInterface is an abstract class for reading data, and each supported format has a dedicated DataInterface that encapsulates the format-specific logic for data reading, metadata extraction, and NWB conversion while presenting a consistent API to users. The DataInterface serves as the fundamental building block of NeuroConv, providing a standardized pathway from diverse source formats to NWB output. This abstraction enables users to work with any supported format using identical code patterns, regardless of the underlying format complexity or internal details. The minimal conversion pipeline is illustrated in {ref}`fig:assets/minimal_conversion_pipeline`:

Reviewer 5: This is a fantastic point and a hallmark of a well-integrated scientific software package. Highlighting the strategy of leveraging and contributing back to the ecosystem is a major strength and a great lesson for the SciPy community.

Author: Thanks a lot for your kind words.

:::{figure} assets/minimal_conversion_pipeline.png
:label: fig:assets/minimal_conversion_pipeline
The process begins with source data (e.g., binary recordings, metadata, and configuration files). A data-specific DataInterface object is instantiated and used to extract metadata via the `interface.get_metadata()` method. The resulting metadata can be optionally edited by the user to fill in missing or corrected fields. The finalized metadata and source data are then passed to the `interface.run_conversion()` method, which writes a complete NWB file compliant with the standard
:::

> 

Reviewer 6: Comments on Figure 1:
1. For consistency,  the title of the third panel could be changed from "Editable" to "Edit metadata".
2. Again in the third panel, "adds" -> "add", "unexisting" -> "nonexistent".
3. In fourth panel, "Adds" -> "Add"
4. The last subtitle could be improved: "Contains… metadata, converted to an NWB-compliant format."
5. Consider leaving out of the caption specific object methods, like get_metadata() etc., and keeping the caption at a high level.

Author: Thanks for all those suggestions. I improved the figures and applied the suggestions.

Programmatically, this process can be summarized in a few lines of code, as shown below. The DataInterface handles all format-specific complexities internally, from parsing proprietary binary structures to extracting embedded metadata, while ensuring the output adheres to NWB best practices:

```python
from datetime import datetime
from zoneinfo import ZoneInfo
from neuroconv import DataInterface

# Initialize the DataInterface for a specific format
interface = DataInterface(file_path="path/to/data/file")

# Extract the metadata from the source format
metadata = interface.get_metadata()

# Modify metadata as needed, add missing or correct existing fields
metadata["NWBFile"]["experimenter"] = ["Baggins, Bilbo"]
metadata["NWBFile"]["experiment_description"] = "Example neurophysiology experiment"
metadata["NWBFile"]["institution"] = "University of Middle Earth"
# session_start_time is required for conversion
metadata["NWBFile"]["session_start_time"] = datetime(2020, 1, 1, 12, 30, 0, tzinfo=ZoneInfo("Middle-earth/Shire"))

# Add the data to an NWB File and write it to disk 
interface.run_conversion(nwbfile_path="path/to/nwbfile.nwb", metadata=metadata)
```

> 

Reviewer 6: Perhaps the second code snippet, which shows the same thing for a specific DataInterface makes the first code snippet redundant.

Author: Yes, it's true that they are somewhat redundant. However, they serve different purposes. The first snippet is meant to be abstract and explain the general workflow, whereas the second is a concrete example from the conversion gallery. Since they play different roles in the flow of the text, I’d prefer to keep both, unless you feel strongly about removing one. In summary, I’d rather be a bit redundant than disrupt the flow here.

This core pattern ensures consistency and simplicity across all supported formats:

1. Instantiate the appropriate DataInterface
2. Extract and modify metadata as needed  
3. Execute the conversion

The DataInterface abstracts away format-specific complexities: from parsing proprietary binary structures to extracting embedded metadata.

Currently supporting 47 distinct input formats (@tbl:formats), each DataInterface is comprehensively documented and demonstrated in the [Conversion Gallery](https://neuroconv.readthedocs.io/en/stable/conversion_examples_gallery/index.html), where users can find complete examples requiring only ~5 lines of code to perform full data conversion. Throughout the conversion process, NeuroConv enforces NWB Best Practices for metadata and data organization while optimizing data storage for both archival purposes and cloud computing requirements.

> 

Reviewer 6: It's great that you link to the conversion gallery. It would also be helpful to link to the index page of the documentation somewhere in the paper (perhaps in the introduction).

Author: Yes, we have a link to the documentation in the first at the beginning of this section: line 55.

As NeuroConv's format support has expanded to 47+ formats, dependency management has become increasingly complex. Each format often requires specialized libraries with potentially conflicting version requirements, creating dependency resolution challenges that can make installation difficult or impossible. For example, libraries for reading different formats may require different versions of Python or different versions of common dependencies such as numpy. Additionally, installing the dependencies of every DataInterface would create an unnecessary and inefficient environment with hundreds of packages, many of which users never need for their conversion.

To address these challenges, we rely on [installation extras](https://packaging.python.org/en/latest/tutorials/installing-packages/#installing-extras) to manage installation complexity. Users can specify only the formats they need during installation:

> 

Reviewer 6: I like this approach to manage dependency conflicts!

## Multi Stream Conversions

Neurophysiology experiments typically involve multiple simultaneous data streams from different modalities, such as raw electrophysiology recordings, spike-sorted data, and behavioral video. Each stream may be stored in a different format, leading to complex conversion requirements. NeuroConv's architecture supports multi-stream conversions through the aggregation of DataInterfaces within a Converter framework as illustrated {ref}`fig:assets/diagram_converter`

:::{figure} assets/diagram_converter.png
:label: fig:assets/diagram_converter
The Converter orchestrates multiple specialized interfaces (A, B, C), each handling different data types (electrophysiology, imaging, and behavior). Individual interfaces extract metadata from their respective data sources, which the Converter combines into a single, user-editable metadata structure through its `converter.get_metadata()` method. The Converter's `converter.run_conversion()` method then coordinates all interfaces to produce a unified NWB file containing all data modalities. This design pattern enables flexible, modular integration of heterogeneous neuroscience data into a single standardized format.

> :::

Reviewer 6: Again, the caption could be more high-level by omitting specific method names. The interfaces in the figure could be named with their proper class names (e.g. OpenEphysRecordingInterface rather than Interface A). Finally, the figure refers to a Converter, but the code in the snippet below and the main text refer to it as ConverterPipe.

Author: The caption was changed to be less specific and more high-level. 

We have a ConverterPipe but that is not the only type of Converter that we have.

Determining optimal chunk parameters involves complex tradeoffs [@zarr_performance]. Large chunks minimize the number of read operations but may require decompressing unnecessary data when accessing small subsets. Small chunks provide more precise access but increase overhead, particularly for cloud storage where each chunk requires a separate HTTP range request. Generally, appropriate chunking requires understanding the most common data access patterns. By understanding common analysis workflows and visualization patterns, it becomes feasible to implement evidence-based heuristics for chunk sizing across common data types, such as voltage recordings and imaging datasets.

> 

Reviewer 6: Is there any available data on how performance scales with chunking? Also, are there benchmarks for how performance scales in high-volume datasets?

Author: Unfortunately not yet that I am aware of, but the community would like to move forward in that direction:

https://nwb-benchmarks.readthedocs.io/en/latest/

# NWB Community and Software Ecosystem

## NeuroConv and the NWB Conversion Landscape

NeuroConv occupies a strategic position within the broader NWB ecosystem, bridging the gap between low-level programming interfaces and high-level user tools. The conversion landscape offers different approaches suited to varying user needs and technical expertise levels.

:::{figure} assets/conversion_comparisons.png
:label: fig:assets/conversion_comparisons

> This illustrates where NeuroConv stands in regard to other conversion tools. For low-level, high-precision control, users can employ the NWB APIs directly ([PyNWB](https://pynwb.readthedocs.io) in Python, [MatNWB](https://matnwb.readthedocs.io/) in MATLAB). For a guided GUI-based experience, the [NWB GUIDE](https://nwb-guide.readthedocs.io/) offers the best option but may be too rigid for complex workflows. NeuroConv stands as a middle ground, automating the conversion of a large number of formats while still allowing for customization and flexibility.

Reviewer 5: This is an excellent and highly informative figure. It clearly positions NeuroConv within the broader ecosystem of NWB conversion tools and helps readers immediately understand its target audience and value proposition. Great inclusion!

Author: Thanks!

Together, NWB, NeuroConv, and DANDI showcase a comprehensive pipeline that spans the entire data lifecycle from acquisition to publication and reuse, supporting emerging software domains from electrophysiological spike sorting to calcium imaging segmentation and behavioral pose estimation.

# Current Limitations

While NeuroConv has significantly improved data standardization processes, some major challenges remain:

**Format Coverage**: Despite supporting 47 formats, the neurophysiology ecosystem continues to evolve with new acquisition systems, format versions, and processing software. Each new system often introduces proprietary formats with unique metadata structures and data organization schemes. While NeuroConv's modular architecture allows users to develop custom DataInterfaces for unsupported formats, this process requires substantial technical expertise in both the source format's internal structure and NeuroConv's interface architecture. Additionally, maintaining custom interfaces as both the source format and NeuroConv evolve presents ongoing maintenance challenges for individual laboratories.

> **Custom Laboratory Formats**: Many laboratories develop custom data storage solutions tailored to their specific experimental workflows, often utilizing MATLAB files, custom CSV schemas, or bespoke binary formats. These custom formats frequently exhibit significant variability not only between laboratories but even within the same laboratory over time as experimental protocols evolve. The heterogeneous nature of these formats, ranging from simple tabular data to complex nested structures with laboratory-specific metadata conventions, makes automated conversion particularly challenging. While NeuroConv can handle some standardized custom formats, the diversity of these approaches often requires manual intervention or custom code development. We recommend that laboratories work with data in its original acquisition format whenever possible, as this provides the richest metadata and most reliable conversion pathway.

Reviewer 5: In the "Custom Laboratory Formats" point, you correctly identify the challenge. You might consider adding a single sentence with a proactive recommendation, such as: "We therefore guide users to develop their custom DataInterface for these formats and offer tutorials and support to facilitate this process." This would turn the limitation into an opportunity for community engagement.

Author: This is a great suggestion. Thanks for pointing this out. In fact, this is an important point that we forget to mention, building a custom data interface is a common workflow when users have behavioral data (which is rarely standardized). I added a sentence here.

# Future Directions

**Enhanced User Experience**: We are implementing comprehensive improvements to examples, tutorials, and documentation. We aim to slowly transition to the Diataxis [@diataxis] framework which will provide clearer learning pathways for users with different backgrounds and goals.

**NWB Standard Evolution**: We actively participate in NWB schema development and maintain close alignment with emerging standards through engagement with NWB Extensions Proposals. By monitoring and contributing to current developments, including enhanced schemas for experimental events (NWBEP001), extracellular electrophysiology (NWBEP002), and optical physiology (NWBEP003, NWBEP004), we ensure that NeuroConv incorporates the latest standard improvements as they become available for the benefit of our users.

> **AI-Assisted Conversion**: We are exploring large language model integration to advance our core mission of automating neurophysiology data conversion. This includes using LLMs to generate DataInterfaces from source format documentation, automatically extract metadata from experimental protocols, and generate custom conversion pipelines based on natural language requirements (conversion agent).

Reviewer 5: This is a very exciting direction! The idea of an LLM-powered "conversion agent" is particularly compelling. To make this more concrete, you could briefly mention the type of LLM interaction you envision (e.g., "fine-tuning a code-generation model on format specifications" or "using RAG with documentation to generate pipeline code"). This isn't required but could add a bit more technical depth to this forward-looking point.

Author: Thanks for this suggestion.

@luiztauffer, could you help me with a sentence in this direction?

Reviewer 2: @h-mayorquin how about this:

> We are developing an LLM-driven conversion agent that uses a lightweight framework (e.g., smolagents) to invoke a specialized suite of NWB-aware tools. Using retrieval-augmented generation, the agent queries a vector store (e.g. Qdrant) populated with neuroconv’s documentation, annotated conversion examples, and NWB best-practice notes, then synthesizes the Python code required for a complete conversion pipeline. Each resulting NWB file is automatically checked with nwbinspector-specialized tool for schema compliance and metadata completeness, while a repository creator tool makes use of our cookie-cutter template to reproduce CI scaffolds, tests, and project structure. By encapsulating years of domain expertise in callable tools and grounding every generation step in verifiable documentation, the agent seeks to streamline converter development, minimize hallucinations, and promote reproducible, standards-conformant data sharing.

Reviewer: Thanks Luiz but I think this is too much for this paper. Let's trim it down to 1-2 sentences and save the rest for its own paper

Author: Thanks again for the suggestion @bhaskargoyal, I added more words in the suggested direction.
